### ISSUE TITLE
ImageLoadingEstimator.LoadImages to load images on demand to avoid OutOfMemoryException

### ISSUE THREAD

**Issue by vpenades** [dotnet/machinelearning#6204]
The problem

ImageLoadingEstimator has two sources:
- LoadImages which loads images from local disk folder on demand.
- LoadRawImageBytes  which takes the images from preloaded bytes from a data column.

The problem is that if you have thousands of images you need to load for training, the only real choice is LoadImages because it will load the images from disk on demand, because LoadRawImageBytes is intended to load just one, or a few images, mostly for inference purposes.

Rephrasing the problem: How to train with thousands of images stored in a network drive accessed by https/ftp/ with credentials?

What I would need:

A way to pass a callback to LoadImages: `delegate Byte[] LoadImageCallback(string imagePathOrUrl);` which would be called on demand.

Or

Use LoadRawImageBytes, but let it accept  Action<Byte[]> as long as Byte[] as the column data, so the memory is allocated on demand.

Without these options, right now I'm completely stucked, because I have my image training data set in a network, so I can't use LoadImages. But I cannot use LoadRawImageBytes either because preloading all those images would throw an OutOfMemoryException.

**Comment by michaelgsharp** [comment]
Hi @vpenades, you should be able to unblock yourself by using the `CustomMapping` transformer. [Here](https://github.com/dotnet/machinelearning/blob/6ae2cff8b99761a3961b6cbfd5f90ad2bf728570/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/CustomMappingSaveAndLoad.cs) is an example that shows how to use the transformer and how to get saving/loading to work (though not with your specific use case).

Basically, you will want to use the `CustomMapping` transformer to load the image bytes and store them in a byte column. You can then use this byte column in conjunction with `LoadRawImageBytes`. This will let you load them however you want, and acts kinda like the `delegate` approach you were mentioning.

Can you try that out and let me know if it works for you? I would be happy to assist if you have further questions or run into more issues. This is in part why we created the `CustomMapping` transformer, because there are going to be needs that we didn't think about, and this will allow many of those needs to be met.

I'll close this issue for now as this should work, but please feel free to reach out/open a new issue if you have further problems.
